### DISCUSSION TITLE
Can't Log Into SSH

### DISCUSSION THREAD

**Discussion by ColtonYYZ** [gtxaspec/wz_mini_hacks#307] [General]
Hi! I don't know what I am doing wrong but when I try to SSH into the cam using this hack, I get a PuTTY Fatal Error. "No supported authentication methods available (server sent: publickey).

I used PuTTYGeN to get my RSA key, trying both 2048 and 4096.

I went into the folder containing the authorized_keys file (\wz_mini_hacks-master\SD_ROOT\wz_mini\etc\ssh). I pasted the key into the file and re-saved it. Then copied all files from the SD_Root folder (wz_mini_hacks-master\SD_ROOT\) on to the root folder of the microSD card and inserted it into the cam. Turned on the cam and tried to SSH into it using the IP and Port 22. 

It first popped up with a message for me to accept a key, and then I got the login prompt "login as:" so I typed "root" without the quotes, and hit enter. That's when I get the fatal error.  

What did I do wrong?

**Comment by sideup66** [comment]
Are you sure your SSH client is using your key? putty is a little weird about the format....has to have hte private key in a .ppk file.

**Comment by ColtonYYZ** [comment]
I did save a PPK file. I thought that was my private key. I saved the public key but there's no extension on that one. When I open it in notepad, I see something that starts off as:  ssh-rsa AAAAB3NzaC1yc2EAAAADAQABAAA

I have tried using PuTTY, I have tried using ssh-keygen in Debian running in Windows Linux Subsystem. I have tried so many ways, and it never works.

This might be a big ask, but could you list the EXACT steps one should take? Talk to me like I am child who doesn't know computers please. 

I have done the steps listed in the GIT:

- perform a fresh format on your micro-sd card, using fat-32 ( this is a hard requirement, the bootloader does not support ex-fat or ext, and thus will not load wz_mini ), DOS partition map type, volume name does not matter.
- copy all the files inside of SD_ROOT to your micro sd card
- SSH is enabled, but is secured using public key authentication for security. Edit the file wz_mini/etc/ssh/authorized_keys and enter your public key here.

I have tried using the guide the GIT points to but I keep getting "Permission Denied" when I try to copy key over, as instructed in the article:

ssh-copy-id -i ~/.ssh/id_rsa.pub username@serveraddress

I have spent hours to no avail.... so if someone would be so kind as to list the steps one would take if they are running Windows 11, with Debian subsystem installed. Hopefully this adds flexibility on which was would be easier.

Question: Accessing SSH via Debian subsystem versus PuTTy, would they require 2 different keys? Or, as soon as I can SSH in (either via Debian or PuTTy), I should be able to do so any other way using the same computer?

**Reply by kohrar** [reply]
When you add your public key (the one that starts off with `ssh-rsa AAA...`) into your `authorized_keys` file, did you use notepad? It's possible the windows carriage return is causing issues. Another issue could be that your PuTTY is too old and doesn't have the right key exchange algorithms.

Since you already have the `id_rsa.pub` file in Debian under WSL, can you copy the contents of this file into the `authorized_keys` file and try again? To avoid messing up the dos/unix formatting, just copy the .pub file over and rename it to authorized_keys on the SD card. Once that's done, try to SSH from your Debian WSL instance.

> Accessing SSH via Debian subsystem versus PuTTy, would they require 2 different keys?

It's up to you. Since PuTTY requires a different format, if you want to use the same key, you'd have to convert your debian one to a PuTTY format (or vice versa). Otherwise, you can just generate two separate keys. Just add each public key in its own line in the `authorized_keys` file.

**Reply by ColtonYYZ** [reply]
I tried simply renaming the file and ssh in but still didn't work. I then decided to stay within Windows and ssh in but failed again!

C:\Users\user\.ssh>copy authorized_keys root@10.0.0.224
        1 file(s) copied.

C:\Users\user\.ssh>ssh root@10.0.0.224
The authenticity of host '10.0.0.224 (10.0.0.224)' can't be established.
ED25519 key fingerprint is SHA256:8/vXxXxXxXxXxX/xXxXxXxXxX+pu2RRI.
This key is not known by any other names
Are you sure you want to continue connecting (yes/no/[fingerprint])? yes
Warning: Permanently added '10.0.0.224' (ED25519) to the list of known hosts.
root@10.0.0.224: Permission denied (publickey).

I've changed the fingerprint listed above to use XxXxXxX only to hide what was shown. 

I dont think I should have said "yes" when it asked if I was sure I wanted to connect.  ED25519 - isn't this different than RSA?

**Reply by kohrar** [reply]
I'm not sure about that copy you did there (you're copying authorized_keys to a file called 'root@10.0.0.224'?) The SSH server's host keys are using ed25519 and isn't relevant to your SSH authorized keys. Just say 'yes' when connecting as you did.

You need to copy the public key file onto your SD card under `SD_ROOT/wz_mini/etc/ssh/authorized_keys`. If you can locate your `id_rsa.pub` file and your SD card is on `E:\`, then you'd do something like `copy id_rsa.pub E:\wz_mini\etc\ssh\authorized_keys`. Eject the SD card, then insert it to the Wyze Cam and then power it on.

**Comment by ColtonYYZ** [comment]
Oh, the part where you say I am copying authorized_keys to a file called 'root@10.0.0.224', this is what the guide states in one of the steps. This guide was provided in section called Setup v3/PANv2 in point # 4.  (https://averagelinuxuser.com/how-to-use-public-key-authentication/)

> How to add SSH keys to a remote computer
> 
> To enable SSH key encryption, you need to place a generated public key to the remote computer. There are different ways to do that. I will show you all the possibilities and you chose the most suitable for you.
> 
> ssh-copy-id
> The simplest one is to use ssh-copy-id:      **ssh-copy-id -i ~/.ssh/id_rsa.pub username@serveraddress**

I thought I could interchange "remote computer" with my WyzeCam with this step.

Also, you say to copy the .pub file to **E:\wz_mini\etc\ssh\authorized_keys** but wouldn't that path be wrong? authorized_keys isn't a folder. It's a file.  So technically that directive path would be 'E:\wz_mini\etc\ssh\authorized_keys\id_rsa.pub'

Or is the example path I just typed be correct and the instructions are wrong?

**Reply by kohrar** [reply]
That's probably where your confusion is coming from. `authorized_keys` is a file and not a directory. You need to put the contents of the public key in this file: https://github.com/gtxaspec/wz_mini_hacks/blob/master/SD_ROOT/wz_mini/etc/ssh/authorized_keys and then copy it onto your SD card.

`ssh-copy-id` won't help you here because you can't log in to the Wyze Cam with a password. You have to set it on the SD card as part of you setting up wz_mini_hacks for the camera. Furthermore, the `copy` command isn't the same as `ssh-copy-id` and can't be interchanged as you've done (you're just going to be copying the file on your computer instead of copying the keys to the device).

**Comment by ColtonYYZ** [comment]
I still can't get it to work. Here are the steps I did from the beginning:

1. I opened command prompt and typed ssh-keygen
2. I did not type any file name and simply pressed "enter" for every step
3. I extracted the zip file downloaded from github
4. I went to the .ssh folder on my windows desktop and COPIED the file and pasted into the same folder as authorized_keys which is found using this path:  /wz_mini/etc/ssh/
5. I copied the name of authorized_keys, deleted that file, and renamed **id_rsa.pub** as **authorized_keys**
6. I used the "safely remove hardware" function in windows and placed the microSD card back into the v3.
7. I heard it say a couple things, ex. extracting...
8. I waited 5 mins and still couldn't view the cam using the Wyze App which I should be able to by that point.
9. I unplugged the cam and re-plugged it in
10. I heard the typically clicking as usual, I could see it in the app, but I still get an error message using PuTTY. See attached
11. I get "Connection closed by 10.0.0.24 port 22 when I use **ssh root@10.0.0.224**

What am I doing wrong?
![PuTTY Error Msg](https://user-images.githubusercontent.com/33189874/191862634-2dfd157b-b4d7-417c-893e-1ab4888185f6.jpg)

**Reply by kohrar** [reply]
I think you did everything correctly in the setup.

The only issue now is that you need to tell PuTTY to use your private keys. Can you locate the id_rsa private key (it's in the same directory as the id_rsa.pub that you copied) and convert it to a putty ppk file?
1. Go to https://www.chiark.greenend.org.uk/~sgtatham/putty/latest.html and download puttygen.exe if you haven't already
2. Run puttygen.exe
3. Go to conversions -> import key. Select your id_rsa file.
4. You should see your key information. Click on 'Save private key'. You can save it without a password. This will generate a /ppk file that PuTTY can use
5. Open PuTTY again. Before you connect, go to the settings under Connection -> SSH -> Auth and specify your private key file. Once that's done, you might want to save this profile so you don't have to do this again.
6. Connect. It should let you log in as root.

**Comment by ColtonYYZ** [comment]
So first off, thank you. We are making progress!  Unfortunately, after carefully following your steps to convert the id_rsa file, I now get "Remote side unexpectedbly closed network connection"

![closed-network-connection](https://user-images.githubusercontent.com/33189874/191865497-2c8a32ab-8bd5-44b1-af9d-2aaf05555440.jpg)

**Reply by kohrar** [reply]
Hmm. I did those exact steps and it works for me (and I don't normally use PuTTY to connect to these Wyze cams)

Are you using the latest version of PuTTY?
Can you try SSH from WSL? (run: ssh -i /path/to/id_rsa root@10.0.0.224) If that doesn't work, add in `-vvv` to your SSH command to get more debug info.

**Comment by ColtonYYZ** [comment]
![Im-In](https://user-images.githubusercontent.com/33189874/191869073-756dc6fb-80a1-4d58-bab3-a99ed71eac2b.jpg)

I removed the microSD card, deleted all the files using my PC, then re-saved the extracted files (which includes the already replaced authorized_keys file I did earlier). I booted up. I heard the audio steps like before... waited this time for 4-5 minutes but it seemed to freeze, so I unplugged it and re-plugged it back in, heard the clicking noises as usual. And VOILA!  I am in!!

And yeah, I am using the newest version of PuTTY. :)

My gosh, 1 cam down, 6 more to go!  LOL   Thank you!!!

![success](https://user-images.githubusercontent.com/33189874/191868741-7383a8a0-3c2e-412a-b9f5-a38f8106a5b1.jpg)
